pkg/load/agents/utils.go

```diff
@@ -62,3 +63,72 @@ func startWatchers(path string, callback func() error, recordError func(string))
 	}
 	return nil
 }
+
+func startReleaseRepoWatcher(path string, callback func() error, recordError func(string)) error {
+	errFunc := func(err error, msg string) {
+		recordError(msg)
+		logrus.WithError(err).Error(msg)
+	}
+
+	watcher, err := GetSymlinkWatcher(callback, errFunc, path)
+	if err != nil {
+		return fmt.Errorf("failed to get the symblink watcher: %w", err)
+	}
+	interrupts.Run(watcher)
+
+	return nil
+}
+
+// GetSymlinkWatcher returns a function that watches if the destination of a single symbolic link changed and run
+// the event function against the new destination.
+func GetSymlinkWatcher(eventFunc func() error, errFunc func(error, string), path string) (func(ctx context.Context), error) {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	oldSymlink, err := os.Readlink(path)
+	if err != nil {
+		return nil, err
+	}
+
+	resolveSymlink := func() string { return filepath.Join(filepath.Dir(path), oldSymlink) }
+
+	if err = w.Add(resolveSymlink()); err != nil {
```

Comment:
I don't think this is what we want.  It will generate an event when the old directory is removed, not when the symbolic link is updated.

Reply:
That's because the operation is done on the directory, not the link.

https://github.com/kubernetes/git-sync/blob/c71d1cb890de054a8be2ee69d023a8c5cf4e1ad1/cmd/git-sync/main.go#L879-L888

```
$ mkdir test; cd test
$ mkdir a b
$ ln -s a l
$ inotifywait --monitor . &
[1] 2768156
$ Setting up watches.
Watches established.

$ ln -snf b tmp
./ CREATE tmp
$ mv -T tmp l
./ MOVED_FROM tmp
./ MOVED_TO l
```

Reply:
Ok, that was introduced in the latest version of `git-sync`. It seems that I was using an older version. Let me check and update the code based on the new behaviour.

Reply:
I think the problem here is how golang is watching the folder. When we start watch the `release` folder which has `rev-123` symlink, `git-sync` creates a new symlink with the new destination `rev-345` as `tmp-link` and then renames the `tmp-link` to `release`, but we still get an event from the old `release` since the new rename event is for `tmp-link` which we don't watch. So if we could just watch the destination folder instead, we can assume that if is deleted, then a new one is created and we can get the new destination name from the new `release` symlink.

Reply:
Yes, which is why we need to monitor the directory:
```go
w, err := fsnotify.NewWatcher()
// …
err := w.Add(".")
// …
for e := range w.Events {
	fmt.Printf("%s\n", e)
}
```

```
$ ./test &
[1] 2683467
$ ln -s a tmp
"./tmp": CREATE
$ mv tmp l
"./tmp": RENAME
"./l": CREATE
$ ln -s b tmp
"./tmp": CREATE
$ mv tmp l
"./tmp": RENAME
"./l": CREATE
$ ln -s a tmp
"./tmp": CREATE
$ mv tmp l
"./tmp": RENAME
"./l": CREATE
```

This is the intended usage of `ConfigMap` mounts, as described [here](https://deploy-preview-233--zealous-shirley-0f56b3.netlify.app/docs/architecture/ci-operator-internals/configuration-updates/#atomicwriter).  It is also how the [`test-infra` agent works](https://github.com/kubernetes/test-infra/blob/07586f2d13d815a4e3053edac7098c36d95a71cb/prow/config/agent.go#L79.).  The recent `entrypoint-wrapper` addition is another [example](https://github.com/openshift/ci-tools/blob/d9519e391da9003ee75b5700e359a8bb9ed7aefb/cmd/entrypoint-wrapper/main.go#L194).

Reply:
I am not sure how this will improve the situation. Why do we need to watch the entire directory? We will still ignore all the events and do the same thing as we do now.

Reply:
It is also safe because we can't rely on reloading the configs on a `CREATE` event for example. We will still rely on the `git-sync` implementation to figure out when is the right time to reload our configs. I will change the code to watch the entire folder instead, but I will need to see the `git-sync` code to determine when we will trigger the reload. I guess on `RENAME`.

Reply:
Not to mention that we will have to implement a new watcher from scratch since the fsnotify won't watch any changes inside the directory.